sdk/core/Azure.Core/src/Pipeline/HttpMessageExtensions.cs

```diff
@@ -0,0 +1,22 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace Azure.Core.Pipeline
+{
+    /// <summary>
+    /// Extension methods for <see cref="HttpMessage"/>.
+    /// </summary>
+    public static class HttpMessageExtensions
+    {
+        /// <summary>
+        /// Sets the package name and version portion of the UserAgent telemetry value.
+        /// Note: If <see cref="DiagnosticsOptions.IsTelemetryEnabled"/> is false, this value is never used.
+        /// </summary>
+        /// <param name="message">The <see cref="HttpMessage"/>.</param>
+        /// <param name="userAgentValue">The <see cref="SetUserAgentString"/>.</param>
+        public static void SetUserAgentString(this HttpMessage message, UserAgentValue userAgentValue)
+        {
+            message.SetInternalProperty(userAgentValue.GetType(), userAgentValue);
```

Comment:
is typeof(UserAgentValue) any faster than calling GetType()?

Reply:
Should we use an internal type as the key so that arbitrary code (with access to the message) cannot set or get the value? Especially that when we read the value (below in the policy), we just retrieve the value from dictionary and call to string. This means that if I have access to the message (through RequestOptions/Context for example), I would be able to do: message.SetProperty(typeof(UserAgentValue), "hello world") and set the header to an arbitrary C# string.

Reply:
> Should we use an internal type as the key so that arbitrary code (with access to the message) cannot set or get the value? Especially that when we read the value (below in the policy), we just retrieve the value from dictionary and call to string. This means that if I have access to the message (through RequestOptions/Context for example), I would be able to do: message.SetProperty(typeof(UserAgentValue), "hello world") and set the header to an arbitrary C# string.

The methods on the message to get and set the type based values are internal, so they could not do this.

Reply:
I suspect we will want to make it (the set property overload) public at some point, at which point we will have to use private key and then it's not clear what the implication of changing the key type are.

Reply:
I think we would be safe to make this change later, but I agree we could do it now as well by introducing a new internal type to act only as the key for a string value whose formatting is controlled by the `UserAgentValue` type.